cypress/integration/unit_tests/color_function_util_test.js

```diff
@@ -10,30 +13,130 @@ let writeToFirebaseStub;
+      byPropertyDiv.append(
+          propertyPicker, doc.createElement('br'),
+          ...makeRadio(doc, 'CONTINUOUS-radio', 'by-property-type'),
+          ...makeRadio(doc, 'DISCRETE-radio', 'by-property-type'),
+          makeTypeDiv(doc, 'continuous'), makeTypeDiv(doc, 'discrete'));
+      colorFunctionEditor.append(makeTypeDiv(doc, 'single'), byPropertyDiv);
+      doc.body.appendChild(colorFunctionEditor);
+    });
   });
 
+  let first = true;
   beforeEach(() => {
+    snackbarStub = cy.stub(snackbar, 'showErrorSnackbar');
+    cy.document().then((doc) => {
+      cy.stub(document, 'getElementById')
+          .callsFake((id) => doc.getElementById(id));
+      // We only need this to run once but stubs are supposed to be set
+      // in beforeEach not before.
+      if (first) {
+        populateColorFunctions();
+        first = false;
+      }
+    });
     writeToFirebaseStub = cy.stub(manageLayersLib, 'updateLayersInFirestore');
-    colorFunctionEditor.hide();
+    $(colorFunctionEditor).hide();
+  });
+
+  it('closes an incomplete color function form', () => {
+    let td = setUpWithLayer({
+      colorFunction: {
+        currentStyle: 0,
+        columns: {
+          wings: {min: 0, max: 100, values: [0, 1, 2, 100]},
+        },
+      },
+    });
+    td.trigger('click');
+    td.trigger('click');
+    expect(snackbarStub.withArgs(
+               'Warning: Closed layer missing color and property. ' +
+               'May not show up on map.'))
+        .to.be.calledOnce;
+
+    const missingColorStub = snackbarStub.withArgs(
+        'Warning: Closed layer missing color. May not show up on map.');
+    td = setUpWithLayer({
+      colorFunction: {
+        currentStyle: 2,
+      },
+    });
+    td.trigger('click');
+    td.trigger('click');
+    expect(missingColorStub).to.be.calledOnce;
+
+    td = setUpWithLayer({
+      colorFunction: {
+        currentStyle: 0,
+        columns: {
+          wings: {min: 0, max: 100, values: [0, 1, 2, 100]},
+        },
+        field: 'wings',
+      },
+    });
+    td.trigger('click');
+    td.trigger('click');
+    expect(missingColorStub).to.be.calledTwice;
+
+    td = setUpWithLayer({
+      colorFunction: {
+        currentStyle: 0,
+        columns: {
+          wings: {min: 0, max: 100, values: [0, 1, 2, 100]},
+        },
+        color: 'red',
+      },
+    });
+    td.trigger('click');
+    td.trigger('click');
+    expect(
+        snackbarStub.withArgs(
+            'Warning: Closed layer missing property. May not show up on map.'))
+        .to.be.calledOnce;
+
+    td = setUpWithLayer({
+      colorFunction: {
+        currentStyle: 1,
+        columns: {
+          wings: {min: 0, max: 100, values: [0, 1, 2, 100]},
+        },
+        field: 'wings',
+        colors: {},
+      },
+    });
+    td.trigger('click');
+    td.trigger('click');
+    expect(snackbarStub.withArgs(
+               'Warning: Closed layer missing at least one color. ' +
```

Comment:
oh, did we want this, or did we want it to warn if *all* colors were missing? @ruthtalbot